### PR TITLE
Improve support for unicode characters

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

In several MLflow APIs, we determine whether or not inputs are of string-like types by performing the following check: `isinstance(item, str)`. Unfortunately, this approach does not treat unicode strings properly.

This PR improves unicode compatibility by modifying these checks to read `isinstance(item, six.string_types)`, which encompasses both `unicode` and `str` types across Python 2 and 3. We already use this improved check in several locations throughout the MLflow API; this PR addresses remaining instances of brittle checks.

## How is this patch tested?

(Details)

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by aarondav — APPROVED** [review]

File: mlflow/projects/_project_spec.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ class Parameter(object):
     """A parameter in an MLproject entry point."""
     def __init__(self, name, yaml_obj):
         self.name = name
-        if isinstance(yaml_obj, str):
+        if isinstance(yaml_obj, six.string_types):
```
Comment (mparkhe) [inline mlflow/projects/_project_spec.py]: Did you consider adding a utility `is_string_type` function?
When python 4 comes out, we prolly dont want to replace `six` with `twelve` all over. Also you can unit test that function with all sort of input -- numbers, unicode, bytes... etc

File: mlflow/utils/validation.py [hunk 2]
```diff
@@ -4,10 +4,12 @@
 import numbers
 import posixpath
 import re
+import six
```
Comment (mparkhe) [inline mlflow/utils/validation.py]: You prolly don’t need this anymore.

File: tests/utils/test_string_utils.py [hunk 3]
```diff
@@ -21,3 +21,16 @@ def test_strip_prefix(original, prefix, expected):
...
+
+def test_is_string_type():
+    assert is_string_type("validstring")
+    assert is_string_type("")
+    assert is_string_type((b'dog').decode("utf-8"))
+    assert not is_string_type(None)
+    assert not is_string_type(["teststring"])
+    assert not is_string_type([])
+    assert not is_string_type({})
+    assert not is_string_type({"test": "string"})
```
Comment (mparkhe) [inline tests/utils/test_string_utils.py]: +1

File: tests/utils/test_string_utils.py [hunk 4]
```diff
@@ -21,3 +21,16 @@ def test_strip_prefix(original, prefix, expected):
 ])
 def test_strip_suffix(original, suffix, expected):
     assert strip_suffix(original, suffix) == expected
+
+
+def test_is_string_type():
```
Comment (mparkhe) [inline tests/utils/test_string_utils.py]: This is great coverage.

**Review by mparkhe — APPROVED** [review]
This is great. Thanks for the awesome cleanup.

**Pull request merged**